# MNIST Classification

### Grid Search for best Hyperparameters

In [54]:
from sklearn.datasets import load_digits
from sklearn.model_selection import GridSearchCV  # helps you chose the model with the best parameters or kernel
from sklearn import svm

In [55]:
import multiprocessing # it is a library that tells you that how many CPU's are available

In [56]:
cpus = multiprocessing.cpu_count()
print(cpus)

12


In [35]:
params = [
    {
        'kernel':['linear','rbf','poly','sigmoid'],
        'C':[0.1,0.2,0.5,1.0,2.0,5.0]
    }
]

## Dataset Prepare

In [42]:
data = load_digits()
X = data.data
Y = ata.target
print(X.shape,Y.shape)

(1797, 64) (1797,)


In [43]:
# normalizing the X data
#u = X.mean()
#std = X.std()
#X = (X-u)/std

## Testing Accuracy For Logistic Regression 

In [44]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score

In [51]:
lr = LogisticRegression(max_iter=5000)
lr.fit(X,Y)
cross_val_score(lr,X,Y,scoring = "accuracy",cv=5).mean()


0.9137604456824512

- So we can see that the Logistic Regression gave an accuracy of about approximately 92%

## Now testing the accuracy for the SVM

In [52]:
from sklearn import svm

In [53]:
svc = svm.SVC()
svc.fit(X,Y)
cross_val_score(svc,X,Y,scoring="accuracy",cv=5).mean()

0.9632838130609718

## Right Set Of Parameters!

In [58]:
gs = GridSearchCV(estimator=svm.SVC(),param_grid = params,scoring="accuracy",cv=5,n_jobs = cpus) # if you will put 'n_jobs = -1' as argument then the function will automatically evaluate the total number of cpu's available for the job

In [59]:
gs.fit(X,Y) # it will train our model 

GridSearchCV(cv=5, error_score=nan,
             estimator=SVC(C=1.0, break_ties=False, cache_size=200,
                           class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='scale', kernel='rbf', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             iid='deprecated', n_jobs=12,
             param_grid=[{'C': [0.1, 0.2, 0.5, 1.0, 2.0, 5.0],
                          'kernel': ['linear', 'rbf', 'poly', 'sigmoid']}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring='accuracy', verbose=0)

In [62]:
# now our model is trained now lets see which is the best estimator ie which model performed the best
gs.best_estimator_  # will specify the best parameters and the best kernel for the data

SVC(C=5.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [63]:
gs.best_score_ # it will specify the best accuracy while having the best paramters and the best accuracy

0.9738502011761063

## SVM with RBF kernel ,C=5.0 gives 97% accuracy on the Dataset,which is the best accuracy that can be achieved with the best kernel and the best values of the required parameters!